File: backend/app/models/message.py
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.database import Base
import uuid

class Message(Base):
    __tablename__ = "messages"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    room_id = Column(UUID(as_uuid=True), ForeignKey("rooms.id", ondelete="CASCADE"))
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    content = Column(Text)
    parent_message_id = Column(UUID(as_uuid=True), ForeignKey("messages.id"))
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    edited_at = Column(DateTime(timezone=True))

    # Relationships
    room = relationship("Room", back_populates="messages")
    user = relationship("User", back_populates="messages")
    parent_message = relationship("Message", remote_side=[id])
    files = relationship("File", back_populates="message", cascade="all, delete-orphan")
    reactions = relationship("MessageReaction", back_populates="message", cascade="all, delete-orphan")

class File(Base):
    __tablename__ = "files"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    message_id = Column(UUID(as_uuid=True), ForeignKey("messages.id", ondelete="CASCADE"), nullable=False)
    file_name = Column(String(255))
    file_size = Column(Integer)
    mime_type = Column(String(100))
    s3_key = Column(String(500))
    thumbnail_s3_key = Column(String(500))
    uploaded_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    message = relationship("Message", back_populates="files")

class MessageReaction(Base):
    __tablename__ = "message_reactions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    message_id = Column(UUID(as_uuid=True), ForeignKey("messages.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    emoji = Column(String(50), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Unique constraint on message_id, user_id, emoji combination
    __table_args__ = (
        {"sqlite_autoincrement": True},
    )

    # Relationships
    message = relationship("Message", back_populates="reactions")
    user = relationship("User", back_populates="reactions")

File: shared/types/message.ts
import { User } from './user'
import { Room } from './room'

export interface Message {
  id: string
  room_id?: string
  user_id: string
  content?: string
  parent_message_id?: string
  created_at: string
  edited_at?: string
  user: User
  room?: Room
  parent_message?: Message
  files?: File[]
  reactions?: MessageReaction[]
}

export interface File {
  id: string
  message_id: string
  file_name?: string
  file_size?: number
  mime_type?: string
  s3_key?: string
  thumbnail_s3_key?: string
  uploaded_at: string
}

export interface MessageReaction {
  id: string
  message_id: string
  user_id: string
  emoji: string
  created_at: string
  user: User
}

export interface CreateMessageRequest {
  room_id?: string
  content?: string
  parent_message_id?: string
  files?: FileUploadInfo[]
}

export interface UpdateMessageRequest {
  content: string
}

export interface FileUploadInfo {
  file_name: string
  file_size: number
  mime_type: string
}

export interface AddReactionRequest {
  emoji: string
}

export interface RemoveReactionRequest {
  emoji: string
}